unify cloudscale/exoscale code

mweibel opened the pull request:
## Summary

* move model packages into reporting package
* get rid of pipeline to improve Go-yness
* adjust & unify logger usage 
* refactor package structure to have less packages with better naming
* refactor writing a record from several implementations into one
* refactor metrics aggregation to all follow the pattern `initialize` -> `accumulate` -> `save`
* unify config for kubernetes env variables
* add integration tests

## Checklist

- [ ] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [ ] Update tests.
- [ ] Link this PR to related issues.

<!--
Remove items that do not apply. For completed items, change [ ] to [x].

NOTE: these things are not required to open a PR and can be done afterwards,
while the PR is open.
-->

glrf commented on pkg/cmd/exoscale.go:
```diff
@@ -2,29 +2,31 @@ package cmd
 	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+// billingHour represents the hour when metrics are collected
+const billingHour = 6
+
```
I checked the old code and it did exactly the same, but could someone explain to me why 6 am and add the explanation in a comment? :sweat_smile:

zugao replied:
It probably has to do with exoscale how they update the data. @Kidswiss probably knows more.

mweibel replied:
I assume the same but I don't know it either. Didn't find it on their website with a very quick search, too. 
Judging by the schedule (`'10 10,16,20 * * *'`) we still crawl it multiple times a day though, just setting the same billing hour to make sure we don't charge too much and just take the latest number.

glrf commented on pkg/reporting/facts.go:
```diff
@@ -1,13 +1,55 @@
+	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+func FactByRecord(ctx context.Context, tx *sqlx.Tx, dt *db.DateTime, record Record) (*db.Fact, error) {
+	query, err := GetQueryByName(ctx, tx, record.QueryName)
```
AFAICT this function is only used in integration tests, right?  In that case I'd prefer to move it to the test suite to reduce API surface.

mweibel replied:
yeah I was wondering where to put it and put it into reporting package to have it at the same location as other Facts related code. I agree about reducing API surface and move it to `test` package :+1:

glrf commented on pkg/reporting/store.go:
```diff
@@ -0,0 +1,168 @@
+	})
+}
+
+// isFactUpdatable makes sure that only missing data or higher quantity values are saved in the billing database
+func isFactUpdatable(ctx context.Context, tx *sqlx.Tx, f *db.Fact, value float64) bool {
+	logger := ctrl.LoggerFrom(ctx)
+
+	fact, _ := GetByFact(ctx, tx, f)
+	if fact == nil || fact.Quantity < value {
+		return true
+	}
+	logger.Info(fmt.Sprintf("skipped saving, higher or equal number already recorded in DB "+
+		"for this hour: saved: \"%v\", new: \"%v\"", fact.Quantity, value))
+	return false
+}
```
Nitpick: AFAIK this functionality only existed for DBaaS before, objectstorage updated the fact unconditionally. But I don't think we intentionally did that before, so this should be fine.

mweibel replied:
yes that's right. Since the check seems to make sense in general, I added it there too. That's one of the only actual thing changing but I really doubt it has any negative impact.

mweibel replied:
For consistency reasons I'd like to keep it - if you agree?

susana-garcia commented on pkg/cmd/exoscale.go:
```diff
@@ -84,22 +83,29 @@ func ExoscaleCmds() *cli.Command {
-					o := sos.NewObjectStorage(exoscaleClient, k8sClient, dbURL)
+					now := time.Now().In(time.UTC)
+					previousDay := now.Day() - 1
+					billingDate := time.Date(now.Year(), now.Month(), previousDay, billingHour, 0, 0, 0, now.Location())
```
not sure if this will work with the first day of the month (`previousDay` calculation)

mweibel replied:
[`time.Date()`](https://pkg.go.dev/time#Date) handles this and other cases well, see: https://go.dev/play/p/Z0k2i8n8Cis

glrf reviewed (approved):
This looks soo much better than before!  

It's hard to review though. I don't see any blocking issues, just very few nitpicks/questions

susana-garcia reviewed (approved):
just a small comment, looks very good.
Also I tested exoscale object bucket with a mysql db locally and it worked perfectly with my old populated db.

zugao reviewed (commented):
I went straight with testing. Unfortunately I had to recreate my whole database which took a while but I think it was worth it. I had no issues with Exoscale DBaaS and ObjectStorage. The story is different with Cloudscale. Connecting to the lab cluster I was not able to save the billing usage of `cloudscale-bucket-lab-test-gsaratura-3` bucket. Also from the logs I could not see any references to that bucket, can you check please?

```2023-01-18T15:44:11.429+0200 | INFO | metrics-collector.cloudscale.objectstorage | cmd/cloudscale.go:81 | Creating cloudscale client
2023-01-18T15:44:11.429+0200 | INFO | metrics-collector.cloudscale.objectstorage | cmd/cloudscale.go:85 | Creating k8s client
2023-01-18T15:44:13.071+0200 | INFO | metrics-collector.cloudscale.objectstorage | cloudscale/objectstorage.go:58 | initialized reporting db
2023-01-18T15:44:17.567+0200 | INFO | metrics-collector.cloudscale.objectstorage | cloudscale/accumulate.go:78 | unable to sync bucket, bucket not found | {"bucket": "testbucket0815"}
2023-01-18T15:44:17.567+0200 | INFO | metrics-collector.cloudscale.objectstorage | cloudscale/accumulate.go:83 | unable to sync bucket, namespace not found | {"namespace": "default"}

Process finished with the exit code 0
```

Also I noticed the DAYS env variable is mandatory. Was that the case already in the old repository?

zugao reviewed (approved):
LGTM

mweibel commented:
clarified the questions via chat and added changes:
- improved logging for cloudscale (issue was different API keys and confusing logging)
- fixed DAYS not being required - nice catch thanks
